logstash-core/src/main/java/org/logstash/instrument/metrics/HistoricMetric.java

```diff
@@ -0,0 +1,233 @@
+ * This directionality allows us to keep track of nodes of a certain age, and to efficiently rotate those links
+ * to new nodes at insertion time. We are tolerant to a reasonable magnitude of out-of-order insertions.
+ *
+ * At insertion time, we maintain links to useful nodes, rotating them forward as needed to avoid full-scale
+ * iteration in either read- or write-modes. These nodes are defined as the _youngest_ node satisfying their
+ * category's minimum time threshold.
+ */
+public class HistoricMetric<T extends Number> {
+
+    // Test Dependency Injection
+    private final Clock clock;
+
+    // Metric Sources
+    private final Metric<T> numeratorMetric;
+    private final Metric<Long> denominatorMetric;
+
+    // Oldest and Newest nodes in linked list
+    private final Node origin;
+    private final AtomicReference<Node> newest;
+
+    // Important intermediate nodes in our linked list
+    private final AtomicReference<Node> quickInstant = new AtomicReference<>();
+    private final AtomicReference<Node> quickOneMinute = new AtomicReference<>();
+    private final AtomicReference<Node> quickFiveMinutes = new AtomicReference<>();
+    private final AtomicReference<Node> quickFifteenMinutes = new AtomicReference<>();
+
+    // Safeguard reads against dead writer
+    private final AtomicReference<Instant> lastRecord = new AtomicReference<>();
+
+    public HistoricMetric(Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this(Clock.systemUTC(), numeratorMetric, denominatorMetric);
+    }
+
+    public HistoricMetric(Metric<T> metric) {
+        this(Clock.systemUTC(), metric);
+    }
+
+    HistoricMetric(Clock clock, Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this.clock = clock;
+        this.numeratorMetric = numeratorMetric;
+        this.denominatorMetric = denominatorMetric;
+
+        // TODO: refactor initial capture out of constructor?
+        this.origin = new Node(clock.instant(), numeratorMetric.getValue(), denominatorMetric.getValue());
+        this.lastRecord.set(this.origin.timestamp);
+        this.newest = new AtomicReference<>(origin);
+    }
+
+    HistoricMetric(Clock clock, Metric<T> metric) {
+        this(clock, metric, new UptimeMetric(clock));
+    }
+
+    // TODO: recording of an HistoricMetric will need to be scheduled
+    public void record() {
+        record(clock.instant(), numeratorMetric.getValue(), denominatorMetric.getValue());
+    }
+
+    void record(final Instant timestamp, final T numerator, final Long denominator) {
+        this.newest.getAndUpdate((previousNewest) -> previousNewest.seekAndPrepend(timestamp, numerator, denominator).getNewest());
+        this.lastRecord.set(timestamp);
+
+        // rotate our quick references in ascending age order, bailing entirely if we are too young
+        if (!rotate(quickInstant, timestamp.minusSeconds(1))) { return; }
```

Comment:
```suggestion
        if (rotate(quickInstant, timestamp.minusSeconds(1)) == false) { return; }
```
Our coming checkstyle starts complaining :)
Also, can we plz declare time values as constant ones?
```
private final static int SECOND = 1;
private final static int MINUTE = 60;
...

```

```suggestion
        if (rotate(quickInstant, timestamp.minusSeconds(SECOND)) == false) { return; }
```